src/plugin/helpers/validation.ts

```diff
@@ -0,0 +1,64 @@
+import type { FieldNameToValueMap } from "../fieldViews/helpers";
+import type { FieldSpec } from "../types/Element";
+
+type Validator = (fieldValue: unknown) => string[];
+
+export const buildValidator = (
+  fieldValidationMap: Record<string, Validator[]>
+) => <FSpec extends FieldSpec<string>>(
+  fieldValues: FieldNameToValueMap<FSpec>
+): Record<string, string[]> => {
+  const errors: Record<string, string[]> = {};
+
+  for (const fieldName in fieldValidationMap) {
+    const validators = fieldValidationMap[fieldName];
+    const value = fieldValues[fieldName];
+    const fieldErrors = validators.flatMap((validator) => validator(value));
+    errors[fieldName] = fieldErrors;
+  }
+  return errors;
+};
+
+export const htmlMaxLength = (maxLength: number): Validator => (value) => {
```

Comment:
It would be good to update the HtmlRequired but as you said given the type issues this is a really good middle ground for us to get the validation ball rolling and working well.